### PR TITLE
add function for tail recursion optimization

### PR DESCRIPTION
**Pull request by pldin601**

### CODE REVIEW

**Comment by pldin601** [comment]
With help of this function you can design your functions to use tail recustion and dont worry about growing the call stack.

```php
$func = tailRec(function ($actual, $max) use (&$func) {
  if ($actual == $max) {
    return $actual;
  }
  return $func($actual + 1, $max);
});

echo $func(1000);
```

This will print `1000`. If you try to call this function without tailRec wrapper you'll get error about nested function limit.

**Comment by lstrojny** [comment]
Sorry for the late response, I think this is a nice addition. Would you mind rebasing?

**Comment by pldin601** [comment]
Hi! Just rebased and merged.

**Comment by lstrojny** [comment]
The rebase doesn’t look right, can you check again?

**Comment by pldin601** [comment]
Tell me please how to rebase correct?

File: src/Functional/TailRecursion.php [hunk 1]
```diff
@@ -9,7 +9,7 @@
  * @param callable $fn
  * @return \Closure
  */
-function tailRec(callable $fn)
+function tailRecursion(callable $fn)
```
Comment (lstrojny) [inline src/Functional/TailRecursion.php]: Sorry, should be `tail_recursion` as we are not using camel case for function names

**Comment by pldin601** [comment]
Renamed.

**Comment by lstrojny** [comment]
Can you remove the superfluous commits from the PR and probably add it to the documentation?

**Comment by pldin601** [comment]
Ok, I will back soon with new PR.

**Pull request closed without merging**